### ISSUE TITLE
Question: EQTransformer model conversion

### ISSUE THREAD

**Issue by calum-chamberlain** [seisbench/seisbench#96]
Hi all,

I wondered if you had considered adding the less-conservative [EQTransformer model](https://github.com/smousavi05/EQTransformer/blob/master/ModelsAndSampleData/EqT_model.h5) from the EQTransformer repo (note that it looks like @smousavi05 has recently renamed the models there) to the pretrained models?

I attempted using the [eqtransformer_conversion](https://github.com/seisbench/seisbench/blob/main/contrib/model_conversion/eqtransformer_conversion.ipynb) notebook but I found that the number of lstm blocks was not the same (two rather than the expected three) as when the notebook was made. It looks like that is the case for both models now and I don't know enough the models well enough to make a judgement on what the correct thing to do is.

Any help or advice would be appreciated on how to convert the models and add the less conservative model, or if it is out of scope and I should figure it out myself then that is fine as well!

**Comment by yetinam** [comment]
Hi Calum,

thanks for the question! To be honest, we have not yet considered adding the other set of EQTransformer weights, mostly because we did not know what the differences between the weights were. If I remember correctly, I decided just to go with the model that was loaded as default in the EQTransformer code. On the other hand, I agree that it would be great to have both weights available.

Another reason we did not include both models in the first place was that they do not fully agree in the architecture (and both differ slightly from the architecture reported in the paper). However, it's pretty straightforward to read out the architecture from the TF models. Similarly, you can adjust quite a few parameters for EQTransformer in the [init call in SeisBench](https://github.com/seisbench/seisbench/blob/1d0b958be686fa45ea1513de2a5ba237153f0a1a/seisbench/models/eqtransformer.py#L37).

Would you be up for converting the new model weights? Maybe they could then be named something like "original2022" and "original2022_conservative". I think the eqtransformer_conversion notebook should be a good starting point. Hopefully it should be enough to only change a few initialisation parameters to the SeisBench EQTransformer call and minor edits to the weight copying. Most of the tf/pytorch problems, e.g., activation in the LSTMs, differences in the norm layers or different padding should already be dealt with. Clearly, if you have any question in the process, let me know and I'll see how I can help.

**Comment by calum-chamberlain** [comment]
I'm up for converting the model weights - I'm working through this now and checking the architectural differences. I'll keep you posted if I need anything, and make a PR with a new contrib notebook for this conversion.

Changes to numpy dependencies between the two projects mean that I haven't been able to run EQTransformer directly in the same environment as seisbench, but I will compare the output for the same test set.

**Comment by calum-chamberlain** [comment]
I haven't made a whole heap of progress unfortunately and could do with some help if and when you have time. I have put my modifications to the notebooks [here](https://github.com/calum-chamberlain/seisbench/blob/eqt-model-conversion/contrib/model_conversion/eqtransformer_model2_conversion.ipynb) and [here for the "conservative" model](https://github.com/calum-chamberlain/seisbench/blob/eqt-model-conversion/contrib/model_conversion/eqtransformer_conservative_conversion.ipynb) in my fork.

It looks like the current model implemented in seisbench is similar to/the same as the [conservative](https://github.com/smousavi05/EQTransformer/raw/master/ModelsAndSampleData/EqT_model_conservative.h5) model. The "original" model has two Bi-LSTM layers instead of three and ends up with a shorter detection window. I get similar results for the "conservative" model between my seisbench conversion and eqtransformer, although not quite as similar as in the conversion deployed [here](https://github.com/seisbench/seisbench/blob/main/contrib/model_conversion/eqtransformer_conversion.ipynb). 

I haven't managed to get a good conversion for the "original" model. You can see the difference between the EQTransformer and my conversion near the bottom of [this notebook](https://github.com/calum-chamberlain/seisbench/blob/eqt-model-conversion/contrib/model_conversion/eqtransformer_model2_conversion.ipynb). I think I'm mapping the convolutions across properly having dropped one Bi-LSTM, but I must have something wrong somewhere! 

Any help appreciated!

**Comment by calum-chamberlain** [comment]
Thanks for that note - changing the LSTM activation made a big difference and the detection and S prediction time series look good now, although the P is still funky. The only other difference in configuration that I noticed was a different droprate between the two models - the other model has a droprate of 0.2 in all the layers that I have seen it listed, but I (naively) assume that this isn't the source of the different predictions.

I may not have renamed the weights correctly for the weights transfer, i will have another look at that tomorrow. 

I implemented a conservative and non-conservative version in `models.eqtransformer` which *might* take care of the differences in architectures. I agree on your suggestion to just update the weights docstring and assign a new version number for the model already included. I think so too given that when I pulled the renamed models in EQTransformer git saw them as being renamed without any other re-writes.

**Comment by yetinam** [comment]
Thanks for the update. I agree, the Dropout is very unlikely to cause the difference. Personally, I find it curious that the P still looks funky, because the same happened in my quick and dirty approach yesterday:

![image](https://user-images.githubusercontent.com/14940857/173798708-6fc7cd5b-fd85-4350-ae95-4789f459349c.png)

Interestingly, the pick seems to be pretty much correct, but then there is this massive blob later. Did you get the same results as I? I find the difference between P and S surprising, because they are literally identical in the SeisBench implementation.

I think the best way for spotting the bug is comparing the outputs layer by layer and see when they start differing. There is some code in the end of the original conversion notebook (`get_all_layer_outputs`) showing how to extract the activations of all intermediate layer from the tensorflow model. For the SeisBench model, I just edited the code to output the activation after the layer of interest (autoreload was extremly helpful here). Plotting the activation maps calculated with tf/pytorch and their difference was quite helpful also to identify the type of issue, e.g., different treatment of the padding.

**Comment by calum-chamberlain** [comment]
Yup, that is pretty much exactly what I see, and I agree that it is odd! I wondered if it was something to do with the attention layers or lstm layers but I didn't find anything obvious. Thanks for the pointer on comparing the layer outputs. I will go through that this afternoon. It is a little frustrating to not being able to use seisbench and EQTransformer in the same environment at the moment, but I will have a go at changing the requirements for EQTransformer to see if I can get them to play nicely.

**Comment by smousavi05** [comment]
Hi All,

Thanks for sharing this thread with me. These observations are very
interesting. I can not think of a reason for that secondary P probability
but can tell this is very weird behavior as haven't seen anything similar
in the thousands of examples I went through.

Mostafa

**Comment by calum-chamberlain** [comment]
Hey @smousavi05 - the secondary P-peak is definitely wrong in my conversion, and is not from your original model! I'm currently looking through individual layer outputs and should at least be able to find out what I have done wrong soon.

**Comment by yetinam** [comment]
Hey @calum-chamberlain,
regarding the environment, I actually do have an environment that runs both EQTransformer and SeisBench. I uploaded the yml for conda [here](https://nextcloud.gfz-potsdam.de/s/NrGfcdc5DXyfLxE). Maybe this helps you. Note that the yml does not include SeisBench, because you'll probably install it as an editable version from the source (`pip install -e .`).

And thanks @smousavi05 for your input! I agree with Calum's and your finding that the secondary P arrival is definitely from a conversion artifact.

**Comment by calum-chamberlain** [comment]
Brief update from my end: On going through all the layers I had to change the encoder copying for the first layer to get similar layer outputs for all layers. For some reason the final result doesn't seem to depend strongly on whether those channels are flipped in order, but there are large differences when comparing intermediate layers. @yetinam - did you see something similar when you were testing your model conversion? The part that I am referring to is the lines:
```python
        # For input layer, resort channels from ENZ to ZNE
        if conv_id == 1:
            data = data[:, [2, 1, 0], :]
```

Once I changed that then the source of the difference seems to be in the P attention layer - the output from the LSTM is the "same" (at least within 1%, I haven't been stricter yet) and the outputs from the detection decoder and S decoder are also the "same", but the P attention layer has quite different results particularly around where the "bump" in the P prediction occurs.

The graph below shows the difference between the output from P attention layer from my conversion and the output from the real EQTransformer:

![image](https://user-images.githubusercontent.com/10913395/174201168-749ec21f-5c61-4da7-bb1e-4b230b4875c0.png)

**Comment by yetinam** [comment]
Thanks for pinpointing the issue to the attention layer. Using this information, I managed to find the issue. For stability in the SelfAttention, there is an epsilon value. However, we used `eps=1e-5`, while the original uses `eps=K.epsilon()` defaulting to `1e-7`. For the other weights, apparently this did not matter too much, but for the weights now, it very much does! I changed the value and the P difference disappeared. Could you verify that this solves the issue for you as well? In that case, I think this will require another small code change for the `original_compatible` models.

For the channel reordering issue, I think you missed the reordering of the inputs when requesting the single layer outputs. There you need to change `wv` to `wv[:, [2, 1, 0]]`. I made this change and then it was okay for me with the channel reordering.

**Comment by calum-chamberlain** [comment]
Awesome, thank you for all the time you have put into helping with this @yetinam - sorry for wasting so much of your time!

Changing `eps=1e-7` fixes that for me as well. I should have spotted that. I will include a change for that when I open a PR for this. The only thing that is left is what looks like a 1-sample shift, which you can see in the original comparison of outputs as well as the spikes at the start and end of the detection window and P and S pick times. Those spikes are small in the conservative model (`~1e-6`), but in the less conservative model they are quite a bit larger (`~0.2`). I'm guessing that this comes from either padding or in the upsampling and I can have a look on Monday.

Before I make a PR I'm planning on running a few tests on other time windows to make sure the results are consistent. Would you like me to include a comparison test as part of the test-set here? If I did that I would probably pre-compute the EQTransformer outputs and include those as files to test against rather than trying to run EQTransformer in the test environment.

Thanks for spotting my channel order mistake as well - I suspected that I had dome something wrong!

**Comment by calum-chamberlain** [comment]
I think I'm happy with the conversion now. I made the change to use `eps=1e-7` which solved most of the issues, but I also found another small bug where the emissions were normalized by the input (x) rather than themselves. I changed the line [here](https://github.com/calum-chamberlain/seisbench/blob/5d68ee35d7893ac6a00fdca074e157b0550e78fd/seisbench/models/eqtransformer.py#L588) to fix that and match the original EQTransformer code [here](https://github.com/smousavi05/EQTransformer/blob/01bf112a00c1a3bbb43adb2c7dba3e21b5fbcdb9/EQTransformer/core/EqT_utils.py#L2512).

@yetinam do you want these changes now, and if so how do you want me to provide the models? 

Or would you rather wait for a comprehensive comparison test between the seisbench implementation and the original EQTransformer? I think that it would be worth getting these small fixes in soon, and it might take me a while to make the comparison tests.

**Comment by yetinam** [comment]
Hi @calum-chamberlain ,
thanks for the update. Great work on finding all the differences! As the model conversion is a lot of tedious manual work and I wouldn't expect to make future changes in this area, I don't see a need to include automatic tests for this. The conversion notebooks for reference should be sufficient.

Regarding the bug, this is really well spotted. Mathematically, it should not make any difference, but it will give numerical differences. Actually, if this bug wouldn't have been there, I'm pretty sure that the `eps` difference wouldn't have caused any issues, because the sum in the divisor is guaranteed to be at least 1 everywhere through the normalisation. Btw the implemented function in the global attention case is anyhow the softmax, so it should be identical (if we didn't code a bug here). Feel free to just delete the lines you commented out there. If necessary in the future, people can still dig them up in the git history.

For the next steps, maybe you could upload the converted weights somewhere and share the link? Then I'll upload them to the repository. For the code changes, just submit a PR against the `maintenance_0.2.x` branch. Once this is merged, I'll create a new patch version so it's available through pypi.

**Comment by calum-chamberlain** [comment]
My concern, and reason for wanting the tests, is that changes to the code, rather than changes to model weights themselves, could introduce differences. But I'm happy to work on those tests elsewhere and post them somewhere as non-automated verification tests that anyone can run for their own sanity.

Good point on softmax - I missed that. To my testing there isn't much difference in time between the `F.softmax` call and running the individual steps, so if you are happy for me to leave it without the `F.softmax` call then I will do.

I will go ahead with that PR and let you know where the weights are. Thanks again for all your help and patience.

**Comment by calum-chamberlain** [comment]
Weights linked and changes made in #101
